src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,608 @@
+                }
+
+                documentUpdate = default;
+                return false;
+            }
+
+            private bool TryConvertIfInToListInvocation(
+                QueryExpressionSyntax source,
+                InvocationExpressionSyntax invocationExpression,
+                out DocumentUpdate documentUpdate)
+            {
+                if (_semanticModel.GetSymbolInfo(invocationExpression, _cancellationToken).Symbol is IMethodSymbol methodSymbol &&
+                    methodSymbol.Parameters.Length == 0)
+                {
+                    return TryConvertIfInInvocation(
+                              source,
+                              invocationExpression,
+                              (listIdentifier, expression) => SyntaxFactory.ExpressionStatement(SyntaxFactory.InvocationExpression(
+                                    SyntaxFactory.MemberAccessExpression(
+                                        SyntaxKind.SimpleMemberAccessExpression,
+                                        listIdentifier,
+                                        SyntaxFactory.IdentifierName(nameof(IList.Add))),
+                                    SyntaxFactory.ArgumentList(SyntaxFactory.SingletonSeparatedList(SyntaxFactory.Argument(expression))))),
+                              SyntaxFactory.ObjectCreationExpression(
+                                  methodSymbol.GenerateReturnTypeSyntax().WithAdditionalAnnotations(Simplifier.Annotation),
+                                  SyntaxFactory.ArgumentList(),
+                                  null),
+                               "list",
+                              out documentUpdate);
+                }
+
+                documentUpdate = default;
+                return false;
+            }
+
+            private bool TryConvertIfInInvocation(
+                QueryExpressionSyntax source,
+                InvocationExpressionSyntax invocationExpression,
+                Func<ExpressionSyntax, ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod,
+                ExpressionSyntax initializer,
+                string variableName,
+                out DocumentUpdate documentUpdate)
+            {
+                var parentStatement = invocationExpression.GetAncestorOrThis<StatementSyntax>();
+                if (parentStatement != null &&
+                    !parentStatement.Equals(default) &&
+                    TryConvertIfInInvocationInternal(
+                        source,
+                        invocationExpression,
+                        parentStatement,
+                        initializer,
+                        variableName,
+                        out var variable,
+                        out var documentCreationMethod) &&
+                      TryConvertQueryExpression(
+                          source,
+                          expression => leafExpressionCreationMethod(variable, expression),
+                          out var statement))
```

Comment:
fwiw, combinign really large expressions like this in an if-condition was super hard to read.  Consider just doing 3 checks, returning early if they fail (or just nest them).